students/FlorentinPopescu/11_Mailroom_Part3_L5_FP.py

```diff
@@ -0,0 +1,187 @@
+                print(f"Error <{os_err}>. Storage directory not indicated. Letters will be saved in the program's directory.\n")  
+            switch_directories(dit, current_directory, new_directory)
+        else:
+            switch_directories(dit, current_directory, new_directory)
+    except FileNotFoundError as fnf_err:
+        print(fnf_err, "; please re-type.")
+        
+#-----------------------------------------------
+def name_donation_letter(dit):
+    while True:
+        entered = input("\nEnter: donor's name; list; quit:>>\n")
+        if entered == 'quit':
+            return None
+        elif entered == 'list':
+            print("\nExisting donors:") 
+            print("".join("\n{}\t{}".format(key, value) for key, value in dit.items()))
+        elif entered not in dit.keys():
+            # gardian - reject names containing special characters, pure numbers, or NaNs. 
+            if check_name(entered) or entered.isdigit() or entered == " ":
+                print("We are sorry, but '{}' does not seem to be a not a valid name. Please re-enter;".format(entered))
+            else:
+                print ("\nAdd new donor: {}".format(entered))
+                try:
+                    amount = input("Welcome! Please enter your generous donation:\n")
+                    amount = float(amount)
+                    #gardian - when donation is positive, add the donor to the list along with donation and print a thank-you note. 
+                    if amount > 0:
+                        add_donation_print_letter(dit, entered, amount)
+                    else: 
+                        print ("We accept only positive donations. Please enter a donation to become one of our donors.")
+                except ValueError as v_err:
+                    print(f"\nWrong amount - error: '{v_err}'. Please re-enter the amount.\n")
+        else: 
+            try:
+                amount = input("Welcome back {}! Please enter your new donation:\n".format(entered))
+                amount = float(amount)
+                if amount > 0:
+                    add_donation_print_letter(dit, entered, amount)
+                #gardian - when an old donor tries to withdraw money inform him his money was spent.
+                elif amount < 0:
+                    print ('Money you donated to charity in the past was spent for noble causes, we cannot return.\n')   
+                else: 
+                    print ("We accept only non-zero donations.")
+            except ValueError as v_err:
+                print(f"\nWrong amount - error: '{v_err}'. Please re-enter the amount.\n")
+                    
+#-----------------------------------------------
+def report(dit):
+    #compute statistics (sum, count, average) 
+    tbl = {}
+    for key, values in dit.items():
+        tbl[key] = [sum(values), len(values), round(sum(values)/len(values), 2)]
+        #tbl.update({key:[sum(values), len(values), round(sum(values)/len(values), 2)]})
```

Comment:
don't see any comprehensions added in the code?

this could be converted into comprehension:
```
stats = {key: [sum(v), len(v), round(sum(v)/len(v), 2)] for k, v in dit.items()}
```